keps/sig-scheduling/20190312-pod-affinity-and-anti-affinity-support-gt-and-lt-operators.md

```diff
@@ -0,0 +1,170 @@
+---
+title: Pod affinity/anti-affinity supports Gt and Lt operators
+authors:
+  - "@wgliang"
+owning-sig: sig-scheduling
+reviewers:
+  - "@bsalamat"
+  - "@k82cn"
+  - "@Huang-Wei"
+approvers:
+  - "@bsalamat"
+  - "@k82cn"
+creation-date: 2019-02-22
+last-updated: 2019-03-12
+status: provisional
+---
+
+ # Pod affinity/anti-affinity supports Gt and Lt operators
+
+ ## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [User Stories](#user-stories)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Design Details](#design-details)
+  * [Content](#content)
+  * [Test Plan](#test-plan)
+  * [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+ ## Summary
+
+ Extend the `Pod` affinity/anti-affinity operators to support `Gt` and `Lt` to provide 
+ users with more elegant Pod label selection capabilities.
+
+ ## Motivation
+
+  We know that `Node` affinity/anti-affinity currently supports `In`, `NotIn`, `Exists`, 
+`DoesNotExist`, `Gt`, `Lt`. But Pod affinity/anti-affinity only works with regular 
+label selectors: `In`, `NotIn`, `Exists`, `DoesNotExist`.
+
+ This is not an ideal situation if users want to put pods based on the label range. 
+ `Pod` affinity/anti-affinity support for `Gt` and `Lt` operators will give users more 
+ control.
+
+ ### Goals
+
+- `Pod` affinity/anti-affinity support for `Gt` and `Lt` operators.
+- `Gt` and `Lt` will have the same status and influence as the original operators (`In`, 
+`NotIn`, `Exists`, `DoesNotExist`).
+
+ ### Non-Goals
+
+- Just add `Gt` and `Lt` operator functions without changing their original definition.
+- Not changing the behavior of other label selectors, such as `Replicases`, `Daemonsets`, etc.
+
+ ## Proposal
+
+ ### User Stories
+
+ As an application developer, I want my application pods to be scheduled onto
+one node that has pod with the "foo" tag and the tag value between "20" and "30".
+
+- if we use the original `In` operator to implement, then we will write affinity 
+like this:
+ ```yaml
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: In
+            values:
+            - 20
+            - 21
+            - 22
+            - 23
+            ...
+            - 28
+            - 29
+            - 30
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+```
+
+ This is not an ideal solution. A promising solution is to provide users with `Gt` and `Lt` 
+ operators, giving users the ability to specify the scope of the tag. Users can achieve 
+ this in this way:
+ ```yaml
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
```

Comment:
Please also mention that this works with `preferredDuringScheduling`(for priorities, soft requirements)..

Reply:
Done.